File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
input = sys.stdin.readline

n = int(input())
arr = [list(map(int, input().split()))for _ in range(n)]
li = [i for i in range(2,n+1)]
start = [1]
link = []
c = sys.maxsize
def dfs():
    if len(start)==n//2:
        a = 0
        b = 0
        global c
        link = list(set(li) - set(start))
        for i in start:
            for j in start:
                a += arr[i-1][j-1]        
        for i in link:
            for j in link:
                b += arr[i-1][j-1]
        c = min(c, abs(a-b))
        return
    
    for i in li:
        if start[-1] < i:
            start.append(i)
            dfs()
            start.pop()

dfs()
print(c)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int[][] graph;
    static boolean[] visited;
    static int ans = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        graph = new int[n + 1][n + 1];
        visited = new boolean[n + 1];

        StringTokenizer st;
        for (int i = 1; i < n + 1; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j < n + 1; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        dfs(0, 0, new ArrayList<>());
        StringBuilder sb = new StringBuilder();
        sb.append(ans);
        System.out.println(ans);
    }

    static void dfs(int start, int cnt, ArrayList<Integer> arr) {
        if (cnt == n / 2) {
            int cal = calculate();
            ans = Math.min(cal, ans);
            return;
        }
        for (int i = start + 1; i < n + 1; i++) {
            visited[i] = true;
            arr.add(i);
            dfs(i, cnt + 1, arr);
            arr.remove(arr.indexOf(i));
            visited[i] = false;
        }
    }

    static int calculate() {
        ArrayList<Integer> start = new ArrayList<>();
        ArrayList<Integer> end = new ArrayList<>();
        int cntStart = 0;
        int cntEnd = 0;

        for (int i = 1; i < visited.length; i++) {
            if (visited[i]) {
                start.add(i);
            } else {
                end.add(i);
            }
        }

        for (Integer i : start) {
            for (Integer integer : start) {
                if (!i.equals(integer)) {
                    cntStart += graph[i][integer];
                }
            }
        }

        for (Integer i : end) {
            for (Integer integer : end) {
                if (!i.equals(integer)) {
                    cntEnd += graph[i][integer];
                }
            }
        }

        return Math.abs(cntStart - cntEnd);
    }
}